server/src/main/java/org/elasticsearch/action/NoShardAvailableActionException.java

```diff
@@ -37,5 +55,31 @@ public RestStatus status() {
 
     public NoShardAvailableActionException(StreamInput in) throws IOException {
         super(in);
+        if (in.getVersion().onOrAfter(Version.V_8_7_0)) {
+            this.onShardFailureWrapper = in.readBoolean();
+        } else {
+            this.onShardFailureWrapper = false;
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        if (out.getVersion().onOrAfter(Version.V_8_7_0)) {
+            out.writeBoolean(this.onShardFailureWrapper);
+        }
+    }
+
+    @Override
+    public StackTraceElement[] getStackTrace() {
+        return onShardFailureWrapper ? EMPTY_STACK_TRACE : super.getStackTrace();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        Map<String, String> delegatedParams = onShardFailureWrapper
```

Comment:
I guess this is important for the case of a coordinating node that is also a data node, which throws this exception and prints it out at REST too, without going through the transport layer. Is it necessary to also forcefully skip the cause?

Reply:
No, we don't need to force to skip the cause or the stacktrace.

Stack trace will be empty from the new field being set to `true` (or from serialization reading in an empty stack trace). The cause is also `null`. 

This code change isn't necessary.